core/src/main/java/org/akaza/openclinica/service/rule/RuleSetService.java

```diff
@@ -852,15 +852,36 @@ private void logExpressions(List<RuleSetBean> validRuleSets) {
 
         for (RuleSetBean ruleSetBean : ruleSets) {
             List<ExpressionBean> expressionsWithCorrectGroupOrdinal = new ArrayList<ExpressionBean>();
+            
+            String itemOid = getExpressionService().getItemOid(ruleSetBean.getExpressions().get(0).getValue());
+            String itemGroupOid = getExpressionService().getItemGroupOid(ruleSetBean.getExpressions().get(0).getValue());
+           
+            HashMap itemDatasHm = getItemDataDao().findByStudyAndOids(ruleSetBean.getStudyId(), itemOid, itemGroupOid);
+            
+            // get the List<ItemDataBean> itemDatas
+            List<ItemDataBean> itemDatas = new ArrayList();
             for (ExpressionBean expression : ruleSetBean.getExpressions()) {
+          	  
                 String studyEventId = getExpressionService().getStudyEventDefinitionOrdninalCurated(expression.getValue());
-                String itemOid = getExpressionService().getItemOid(expression.getValue());
-                String itemGroupOid = getExpressionService().getItemGroupOid(expression.getValue());
+                itemOid = getExpressionService().getItemOid(expression.getValue());
+                itemGroupOid = getExpressionService().getItemGroupOid(expression.getValue());
                 String groupOrdinal = getExpressionService().getGroupOrdninalCurated(expression.getValue());
-                List<ItemDataBean> itemDatas = getItemDataDao().findByStudyEventAndOids(Integer.valueOf(studyEventId), itemOid, itemGroupOid);
+              
                 logger.debug("studyEventId {} , itemOid {} , itemGroupOid {} , groupOrdinal {} , itemDatas {}", new Object[] { studyEventId, itemOid,
                     itemGroupOid, groupOrdinal, itemDatas.size() });
 
+                String key = ruleSetBean.getStudyId() + studyEventId + itemGroupOid +itemOid;
+                
+                if(itemDatasHm.containsKey(key)) {
+                	itemDatas.add((ItemDataBean) itemDatasHm.get(key));
```

Comment:
Looks like we are using the "itemDatas" variable only to get the count of item data. Can we just write a count query instead of loading all item data into a hashmap?

Reply:
Done - write a count query instead of loading all item data into a hashmap